perf(allocator/vec): remove `SetLenOnDrop`

overlookmotel opened the pull request:
Simplify `Vec::truncate` and `Vec::extend_with` methods.

Unlike standard library, our `Vec<'a, T>` requires that `T` is not `Drop`. Therefore we don't need to worry about dropping elements.

There are likely many more places where we can make similar simplifications (optimizations). I've only done these 2 now in order to get rid of `SetLenOnDrop`. I wanted to get rid of it because it requires direct access to `self.buf.len`, which I'd like to make private in a later PR.

overlookmotel commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/11079?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#11079** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/11079?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/11079?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#11074** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/11074?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#11073** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/11073?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#11072** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/11072?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#11071** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/11071?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=11079).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

overlookmotel commented on crates/oxc_allocator/src/vec2/mod.rs:
```diff
@@ -2081,66 +2072,34 @@ impl<'bump, T: 'bump> Vec<'bump, T> {
-            // len set by scope guard
+            // `n` and `self.len() + n` must both be `<= u32::MAX`, otherwise `self.reserve(n)` above
+            // would have panicked. So `n as u32` cannot truncate `n`, and `len + n` cannot wrap.
+            #[expect(clippy::cast_possible_truncation)]
```
Oh do shut up about `#[expect]`, you doofus!

Dunqing commented on crates/oxc_allocator/src/vec2/mod.rs:
```diff
@@ -2081,66 +2072,34 @@ impl<'bump, T: 'bump> Vec<'bump, T> {
     fn extend_with<E: ExtendWith<T>>(&mut self, n: usize, mut value: E) {
         self.reserve(n);
 
+        // std library version of this method updates `len` after writing each element,
+        // because `value.next()` or `value.last()` could panic.
+        // If that happens, `len` needs to contain all the elements written so far,
+        // so they get dropped when the `Vec` is dropped.
+        // But our `Vec` requires that `T` is not `Drop`, so we don't need to worry about that.
```
What does this mean? We don't need to worry about the `len` being incorrect, or is it due we don't have a way to deal with the `SetLenOnDrop::drop` thing, so we don't need to worry about it?

overlookmotel replied:
Oh damn it! The way I wrote this comment is obviously not very clear.

### What stdlib does

stdlib's `Vec` allows the type it contains `T` to be `Drop`. When the `Vec` is dropped, it must drop the individual `T`s it contains first. This is performed in the `Drop` impl for `Vec`.

`Vec::resize` takes a value and clones it multiple times, inserting each clone into the `Vec`.

The problem is that `Clone::clone` can panic (because it can be user-defined).

Let's consider what happens if `clone` panics on the 2nd call if you don't set `len` after each turn of the loop:

* Our `Vec` is a `Vec<String>`.
* Initial length of `Vec` is 0.

1. `String` is cloned and added to the `Vec`.
2. `len` of the `Vec` is still 0 (`len` is only incremented in a local var inside `extend_with`).
3. `extend_with` tries to clone a 2nd value, but `clone` panics.
4. Stack unwinds, and all objects on stack are dropped.
5. The `Vec` is dropped.
6. `Vec::drop` must drop all the `String`s it owns.
7. It checks it's `len` field, but it's 0, so it concludes "I don't own anything, nothing to do".
8. The first `String` which was inserted into the `Vec` in step 2 is leaked -> memory leak.

So the solution std lib takes is to always set the `len` field after each element is added to the `Vec`. `len` is always up to date. This ensures that if there's a panic, `Vec::drop` knows exactly how many elements it owns, and will drop them all.

It does this via the `SetLenOnDrop` structure, which itself has a `Drop` impl which sets the `len` field when it's dropped. It's pretty complicated!

### Why we can do it simpler

In our case, we require that `T` in `Vec<T>` is not `Drop`, and `Vec` itself has no `Drop` implementation. So nothing ever needs to be dropped. That's what I mean about "we don't need to worry about it".

Consequently all the complication with `SetLenOnDrop` is unnecessary, and we can just set the `len` field once at the end. If `clone` panics, then `len` field will be left in an inaccurate state, but that doesn't matter as nothing reads it anyway.

Dunqing replied:
Thank you for taking the time to answer my question!

> Oh damn it! The way I wrote this comment is obviously not very clear.

Not at all, the comment is good! I don't understand due to my lack of Rust low-level knowledge.

Dunqing reviewed (approved):
Looks good! Your comments let me know better what `SetLenOnDrop` is going on.

overlookmotel commented:
### Merge activity

* **May 17, 6:15 AM EDT**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

overlookmotel commented:
I'll fix the Miri fail. It appears not to be due to Miri flagging UB, but due to lint errors which have been added on nightly.

overlookmotel commented:
Mire fail was unrelated. Fixed in #11092.

overlookmotel committed:
perf(allocator/vec): remove `SetLenOnDrop` (#11079)

Simplify `Vec::truncate` and `Vec::extend_with` methods.

Unlike standard library, our `Vec<'a, T>` requires that `T` is not `Drop`. Therefore we don't need to worry about dropping elements.

There are likely many more places where we can make similar simplifications (optimizations). I've only done these 2 now in order to get rid of `SetLenOnDrop`. I wanted to get rid of it because it requires direct access to `self.buf.len`, which I'd like to make private in a later PR.
